MacOS Fails to build on boost@1.86

I started running into this issue today when building esbmc.

> `src/util/crypto_hash.cpp:47:26: error: non-const lvalue reference to type 'digest_type' (aka 'unsigned char[20]') cannot bind to a value of unrelated type 'unsigned int[5]'
>   p_crypto->s.get_digest(hash);



> /opt/homebrew/include/boost/uuid/detail/sha1.hpp:179:43: note: passing argument to parameter 'digest' here
> inline void sha1::get_digest(digest_type& digest)`

On investigating, it seems by boost version was upgraded from 1.85 to 1.86.
The change log has this particular line of code from 1.85 to 1.86
`typedef unsigned int(digest_type)[5]; => typedef unsigned char digest_type[ 20 ];`

reverting to boost 1.85 works.

--- a/src/util/crypto_hash.cpp
+++ b/src/util/crypto_hash.cpp
@@ -26,8 +26,15 @@ bool crypto_hash::operator<(const crypto_hash &h2) const
 std::string crypto_hash::to_string() const
 {
   std::ostringstream buf;
-  for (unsigned int i : hash)
+#if BOOST_VERSION >= 108600
+  for (HashType c : hash)
+  {
+    buf << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(c);
+  }
+#else
+  for (HashType i : hash)
     buf << std::hex << std::setfill('0') << std::setw(8) << i;
+#endif
 
   return buf.str();
 }

--- a/src/util/crypto_hash.h
+++ b/src/util/crypto_hash.h
@@ -4,20 +4,28 @@
 #include <memory>
 #include <string>
 
+#if BOOST_VERSION >= 108600
+#  define HASH_SIZE 20
+using HashType = unsigned char;
+#else
+#  define HASH_SIZE 5
+using HashType = unsigned int;
+#endif
+
 class crypto_hash_private;
 
 class crypto_hash
 {
 public:
   std::shared_ptr<crypto_hash_private> p_crypto;
-  unsigned int hash[5];
+  HashType hash[HASH_SIZE];
 
   bool operator<(const crypto_hash &h2) const;
 
   size_t to_size_t() const
   {
     size_t result = hash[0];
-    for (int i = 1; i < 5; i++)
+    for (int i = 1; i < HASH_SIZE; i++)
       // Do we care about overlaps?
       result ^= hash[i];
     return result;

